{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a585a329-da0e-4708-901c-e9800cc8cdcd",
   "metadata": {},
   "source": [
    "# Analysis for 9p21.3 Modifier Screen\n",
    "\n",
    "EnAsCas12-expressing KP4 cells were transduced with a dCas9-KRAB targeting either Ch2-2 or PELO (gRNA #1/2), then a Cas12a library targeting 9p21.3 genes (and positive/negative controls) was applied.  The initial pDNA pool was sequenced as a day 0 reference, with additional timepoints being taken at day 5, 9, 13, 17 and 21.  In this notebook, we use raw readcounts to compute naive logfold change (LFC) as well as run Chronos to obtain gene effect scores.  \n",
    "\n",
    "NOTE: Chronos is non-deterministic, so its outputs will minorly fluctuate from those presented in the manuscript."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d658f818-774c-4112-bc7f-5eb8350f682a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from chronos.model import normalize_readcounts, Chronos, calculate_fold_change\n",
    "from helper_funcs import *"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a621ee4e-96a6-4066-9c5c-7ef54f1bfcde",
   "metadata": {},
   "source": [
    "# Import data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "32bb5a03-4ed1-454e-b4bf-6f478609dd29",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetching local copy of readcounts_9p21_3\n",
      "Fetching local copy of guide_map_9p21_3\n",
      "Fetching local copy of sequence_map_9p21_3\n"
     ]
    }
   ],
   "source": [
    "redownload = False\n",
    "reads = read_in_data(\"readcounts_9p21_3\", index_col=0, redownload=redownload)\n",
    "guide_map = read_in_data(\"guide_map_9p21_3\", redownload=redownload).rename(columns={\"sgRNA1\":\"sgrna\", \"Target\":\"gene\"})\n",
    "seq_map = read_in_data(\"sequence_map_9p21_3\", redownload=redownload)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c65d1fc4-c8cc-495b-aa37-5f214397c14d",
   "metadata": {},
   "source": [
    "# Remove multitargeting guides"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c143c956-5f42-4b08-bd10-2e33015674db",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['ACTGAACTTTACCAGCAACTGAA', 'CCATTTGTGCCAGGAGTATCAAG',\n",
      "       'TCTGGGCTGTGATCTGCCTCAGA'],\n",
      "      dtype='object', name='sgrna')\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(92, 46)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "multitargeting = guide_map.groupby('sgrna')['gene'].count().loc[lambda x: x > 1].index\n",
    "print(multitargeting)\n",
    "guide_map = guide_map.loc[~guide_map[\"sgrna\"].isin(multitargeting)]\n",
    "reads = reads.loc[~reads.index.isin(multitargeting)]\n",
    "reads.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98057dac-1135-4d88-bc49-74fa463c68df",
   "metadata": {},
   "source": [
    "# Compute LFCs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6eddddf4-cb75-4a6b-93df-130fc82348ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "lfc = np.log2(calculate_fold_change(reads.T, seq_map))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1e83e4e-7029-4af5-99ee-41c9822790a6",
   "metadata": {},
   "source": [
    "# Train Chronos\n",
    "\n",
    "Due to small scale of this library, kernel_width and cell_efficacy_guide_quantile were tweaked from their default whole-genome settings."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "909affc6-1653-4a03-8cb5-01158c4b0594",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "normalizing readcounts\n",
      "Readcounts has less than 2000 guides, using median normalization\n",
      "\n",
      "\n",
      "Finding all unique guides and genes\n",
      "found 92 unique guides and 27 unique genes in 9p21.3-genes\n",
      "found 92 unique guides and 27 unique genes overall\n",
      "\n",
      "finding guide-gene mapping indices\n",
      "\n",
      "finding all unique sequenced replicates, cell lines, and pDNA batches\n",
      "found 45 unique sequences (excluding pDNA) and 3 unique cell lines in 9p21.3-genes\n",
      "found 45 unique replicates and 3 unique cell lines overall\n",
      "\n",
      "finding replicate-cell line mappings indices\n",
      "\n",
      "finding replicate-pDNA mappings indices\n",
      "\n",
      "\n",
      "assigning float constants\n",
      "Estimating or aligning variances\n",
      "Creating excess variance tensors\n",
      "\tCreated excess variance tensor for 9p21.3-genes with shape [45, 1]\n",
      "initializing graph\n",
      "\n",
      "building gene effect mask\n",
      "\n",
      "building doubling vectors\n",
      "made days vector of shape [45, 1] for 9p21.3-genes\n",
      "\n",
      "building late observed timepoints\n",
      "\tbuilt normalized timepoints for 9p21.3-genes with shape (45, 92) (replicates X guides)\n",
      "\n",
      "building t0 reads\n",
      "\n",
      "\n",
      "Building variables\n",
      "\n",
      "building t0 reads estimate\n",
      "made t0 batch with shape (1, 92) for 9p21.3-genes\n",
      "building guide efficacy\n",
      "built guide efficacy: shape [1, 93]\n",
      "building growth rate\n",
      "built growth rate: shape {'9p21.3-genes': [3, 1]}\n",
      "\n",
      "building cell line efficacy\n",
      "built cell line efficacy: shapes {'9p21.3-genes': [3, 1]}\n",
      "building screen delay\n",
      "built screen delay\n",
      "building gene effect\n",
      "built core gene effect: 3 cell lines by 27 genes\n",
      "\n",
      "\n",
      "Connecting graph nodes in model\n",
      "\n",
      "building effective days\n",
      "built effective days, shapes {'9p21.3-genes': [45, 27]}\n",
      "\n",
      "building gene effect growth graph nodes\n",
      "built gene effect growth graph nodes, shapes {'9p21.3-genes': [3, 27]}\n",
      "\n",
      "building combined efficacy\n",
      "built combined efficacy, shape {'9p21.3-genes': [3, 92]}\n",
      "built expanded combined efficacy, shapes {'9p21.3-genes': [45, 92]}\n",
      "\n",
      "building growth estimates of edited cells and overall estimates of fold change in guide abundance\n",
      "built growth and change\n",
      "\n",
      "building unnormalized estimates of final abundance\n",
      "built unnormalized abundance\n",
      "\n",
      "building normalized estimates of final abundance\n",
      "built normalized abundance\n",
      "\n",
      "\n",
      "Building all costs\n",
      "\n",
      "assembling guide efficacy regularization\n",
      "building smoothed regularization\n",
      "\n",
      "building t0 reads regularization/cost\n",
      "\n",
      "building NB2 cost\n",
      "building other regularizations\n",
      "\n",
      "Creating optimizer\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-10-18 15:41:59.475450: I tensorflow/compiler/mlir/mlir_graph_optimization_pass.cc:388] MLIR V1 optimization pass is not enabled\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "initializing variables\n",
      "estimating initial screen efficacy and gene effect\n",
      "\t 9p21.3-genes\n",
      "\tverifying graph integrity\n",
      "verifying user inputs\n",
      "verifying variables\n",
      "verifying calculated terms\n",
      "\t9p21.3-genes _gene_effect\n",
      "\t9p21.3-genes _selected_efficacies\n",
      "\t9p21.3-genes_predicted_readcounts_unscaled\n",
      "\t9p21.3-genes _predicted_readcounts\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-10-18 15:41:59.799731: W tensorflow/c/c_api.cc:305] Operation '{name:'excess_variance/9p21.3-genes/Assign' id:6 op device:{requested: '', assigned: ''} def:{{{node excess_variance/9p21.3-genes/Assign}} = AssignVariableOp[_has_manual_control_dependencies=true, dtype=DT_DOUBLE, validate_shape=false](excess_variance/9p21.3-genes, excess_variance/9p21.3-genes/Initializer/initial_value)}}' was changed by setting attribute after it was run by a session. This mutation will have no effect, and will trigger an error in the future. Either don't modify nodes after running them or create a new session.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\t9p21.3-genes _normalized_readcounts\n",
      "\t9p21.3-genes _cost_presum\n",
      "sess run\n",
      "\t9p21.3-genes _cost\n",
      "\t9p21.3-genes _full_costs\n",
      "ready to train\n",
      "NB2 cost 0.4140102137294511\n",
      "Full cost 0.41598662502482053\n",
      "relative_growth_rate\n",
      "\t9p21.3-genes max 1.027, min 0.98626\n",
      "mean guide efficacy 0.990853919829184\n",
      "t0_offset SD: [('9p21.3-genes', 7.477999527876274e-05)]\n",
      "\n",
      "gene mean 0.0017752704167814711\n",
      "SD of gene means 0.3417455449072775\n",
      "Mean of gene SDs 0.05221446186476038\n",
      "\n",
      "\n",
      "\n",
      "51 epochs trained, time taken 0:00:00, projected remaining 0:00:01\n",
      "NB2 cost 0.24770009478089539\n",
      "Full cost 0.26738688934908655\n",
      "relative_growth_rate\n",
      "\t9p21.3-genes max 1.057, min 0.91008\n",
      "mean guide efficacy 0.9600038241879154\n",
      "t0_offset SD: [('9p21.3-genes', 0.13501652635274505)]\n",
      "\n",
      "gene mean -0.0035153813355761417\n",
      "SD of gene means 0.6038182107386302\n",
      "Mean of gene SDs 0.07434223062749276\n",
      "\n",
      "\n",
      "\n",
      "101 epochs trained, time taken 0:00:00, projected remaining 0:00:01\n",
      "NB2 cost 0.2083492957086144\n",
      "Full cost 0.22163034032535395\n",
      "relative_growth_rate\n",
      "\t9p21.3-genes max 1.049, min 0.96399\n",
      "mean guide efficacy 0.8717513264933451\n",
      "t0_offset SD: [('9p21.3-genes', 0.10664230802480867)]\n",
      "\n",
      "gene mean -0.0016698588613301116\n",
      "SD of gene means 0.9901711879580386\n",
      "Mean of gene SDs 0.05505636908130593\n",
      "\n",
      "\n",
      "\n",
      "151 epochs trained, time taken 0:00:00, projected remaining 0:00:00\n",
      "NB2 cost 0.20382903276414316\n",
      "Full cost 0.21583987091538678\n",
      "relative_growth_rate\n",
      "\t9p21.3-genes max 1.049, min 0.97076\n",
      "mean guide efficacy 0.8739919104295247\n",
      "t0_offset SD: [('9p21.3-genes', 0.09890899390017016)]\n",
      "\n",
      "gene mean -0.05569965886465961\n",
      "SD of gene means 1.1144411834606898\n",
      "Mean of gene SDs 0.055431099949408216\n",
      "\n",
      "\n",
      "\n",
      "201 epochs trained, time taken 0:00:00, projected remaining 0:00:00\n",
      "NB2 cost 0.20531239821786562\n",
      "Full cost 0.21883182115509195\n",
      "relative_growth_rate\n",
      "\t9p21.3-genes max 1.046, min 0.97686\n",
      "mean guide efficacy 0.861096541951124\n",
      "t0_offset SD: [('9p21.3-genes', 0.10344690617070089)]\n",
      "\n",
      "gene mean -0.09072614186881228\n",
      "SD of gene means 1.187859932038488\n",
      "Mean of gene SDs 0.05442327941017798\n",
      "\n",
      "\n",
      "\n",
      "251 epochs trained, time taken 0:00:00, projected remaining 0:00:00\n",
      "NB2 cost 0.20365559152906423\n",
      "Full cost 0.2164548383498129\n",
      "relative_growth_rate\n",
      "\t9p21.3-genes max 1.047, min 0.97455\n",
      "mean guide efficacy 0.8755366654764759\n",
      "t0_offset SD: [('9p21.3-genes', 0.09973568153697884)]\n",
      "\n",
      "gene mean -0.09859536294078006\n",
      "SD of gene means 1.241916853463002\n",
      "Mean of gene SDs 0.05417520554092823\n",
      "\n",
      "\n",
      "\n",
      "301 epochs trained, time taken 0:00:01, projected remaining 0:00:00\n",
      "NB2 cost 0.20168205311906443\n",
      "Full cost 0.2147911953357258\n",
      "relative_growth_rate\n",
      "\t9p21.3-genes max 1.046, min 0.97605\n",
      "mean guide efficacy 0.8622829060176314\n",
      "t0_offset SD: [('9p21.3-genes', 0.09933229941041466)]\n",
      "\n",
      "gene mean -0.11067626141351307\n",
      "SD of gene means 1.2765691865565623\n",
      "Mean of gene SDs 0.052795157229137135\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "model = Chronos(\n",
    "    sequence_map={\"9p21.3-genes\": seq_map},\n",
    "    guide_gene_map={\"9p21.3-genes\": guide_map},\n",
    "    readcounts={\"9p21.3-genes\": reads.T},\n",
    "    kernel_width=3,\n",
    "    cell_efficacy_guide_quantile=0.15\n",
    ")\n",
    "model.train()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7fbf8b43-5c54-47e6-b7c8-e9645f9aff40",
   "metadata": {},
   "source": [
    "# Scale gene effect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "eb03eeee-63a5-4b00-829b-1e3e92b9a8d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "neg_con = guide_map.loc[guide_map['Function'] == 'Negative Control', 'gene'].unique()\n",
    "pos_con = guide_map.loc[guide_map['Function'] == 'Lethal Control', 'gene'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "1d8c5904-859e-4c65-8914-6296bf653724",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>gene</th>\n",
       "      <th>AAVS</th>\n",
       "      <th>BLM</th>\n",
       "      <th>CD47</th>\n",
       "      <th>CD63</th>\n",
       "      <th>CDKN2A</th>\n",
       "      <th>Ch22</th>\n",
       "      <th>FOCAD</th>\n",
       "      <th>HACD4</th>\n",
       "      <th>IFNA10</th>\n",
       "      <th>IFNA14</th>\n",
       "      <th>...</th>\n",
       "      <th>POLR1C</th>\n",
       "      <th>POLR2D</th>\n",
       "      <th>RAD51</th>\n",
       "      <th>RPA1</th>\n",
       "      <th>RPA2</th>\n",
       "      <th>RecQL1</th>\n",
       "      <th>RecQL4</th>\n",
       "      <th>RecQL5</th>\n",
       "      <th>SF3B1</th>\n",
       "      <th>WRN</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Ch2-2</th>\n",
       "      <td>0.198210</td>\n",
       "      <td>-0.264190</td>\n",
       "      <td>-0.019702</td>\n",
       "      <td>-0.202424</td>\n",
       "      <td>0.161687</td>\n",
       "      <td>0.086504</td>\n",
       "      <td>-0.159614</td>\n",
       "      <td>-0.115506</td>\n",
       "      <td>0.030651</td>\n",
       "      <td>-0.073329</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.109926</td>\n",
       "      <td>-1.337034</td>\n",
       "      <td>-1.099350</td>\n",
       "      <td>-1.598327</td>\n",
       "      <td>-0.746743</td>\n",
       "      <td>0.005388</td>\n",
       "      <td>-0.209613</td>\n",
       "      <td>0.024706</td>\n",
       "      <td>-1.357318</td>\n",
       "      <td>0.022348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PELO1</th>\n",
       "      <td>0.210632</td>\n",
       "      <td>-0.260169</td>\n",
       "      <td>-0.033406</td>\n",
       "      <td>-0.183425</td>\n",
       "      <td>0.161934</td>\n",
       "      <td>0.054452</td>\n",
       "      <td>-0.478578</td>\n",
       "      <td>-0.090828</td>\n",
       "      <td>0.034737</td>\n",
       "      <td>-0.056125</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.036756</td>\n",
       "      <td>-1.310390</td>\n",
       "      <td>-1.109442</td>\n",
       "      <td>-1.613100</td>\n",
       "      <td>-0.682663</td>\n",
       "      <td>-0.000709</td>\n",
       "      <td>-0.138383</td>\n",
       "      <td>0.043668</td>\n",
       "      <td>-1.343628</td>\n",
       "      <td>0.043510</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PELO2</th>\n",
       "      <td>0.219337</td>\n",
       "      <td>-0.250749</td>\n",
       "      <td>-0.072253</td>\n",
       "      <td>-0.192066</td>\n",
       "      <td>0.160570</td>\n",
       "      <td>0.063674</td>\n",
       "      <td>-0.541295</td>\n",
       "      <td>-0.075546</td>\n",
       "      <td>0.036401</td>\n",
       "      <td>-0.055231</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.044134</td>\n",
       "      <td>-1.273908</td>\n",
       "      <td>-1.052479</td>\n",
       "      <td>-1.624128</td>\n",
       "      <td>-0.686925</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.145139</td>\n",
       "      <td>0.042951</td>\n",
       "      <td>-1.361224</td>\n",
       "      <td>0.054788</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "gene                AAVS       BLM      CD47      CD63    CDKN2A      Ch22  \\\n",
       "cell_line_name                                                               \n",
       "Ch2-2           0.198210 -0.264190 -0.019702 -0.202424  0.161687  0.086504   \n",
       "PELO1           0.210632 -0.260169 -0.033406 -0.183425  0.161934  0.054452   \n",
       "PELO2           0.219337 -0.250749 -0.072253 -0.192066  0.160570  0.063674   \n",
       "\n",
       "gene               FOCAD     HACD4    IFNA10    IFNA14  ...    POLR1C  \\\n",
       "cell_line_name                                          ...             \n",
       "Ch2-2          -0.159614 -0.115506  0.030651 -0.073329  ... -1.109926   \n",
       "PELO1          -0.478578 -0.090828  0.034737 -0.056125  ... -1.036756   \n",
       "PELO2          -0.541295 -0.075546  0.036401 -0.055231  ... -1.044134   \n",
       "\n",
       "gene              POLR2D     RAD51      RPA1      RPA2    RecQL1    RecQL4  \\\n",
       "cell_line_name                                                               \n",
       "Ch2-2          -1.337034 -1.099350 -1.598327 -0.746743  0.005388 -0.209613   \n",
       "PELO1          -1.310390 -1.109442 -1.613100 -0.682663 -0.000709 -0.138383   \n",
       "PELO2          -1.273908 -1.052479 -1.624128 -0.686925  0.000000 -0.145139   \n",
       "\n",
       "gene              RecQL5     SF3B1       WRN  \n",
       "cell_line_name                                \n",
       "Ch2-2           0.024706 -1.357318  0.022348  \n",
       "PELO1           0.043668 -1.343628  0.043510  \n",
       "PELO2           0.042951 -1.361224  0.054788  \n",
       "\n",
       "[3 rows x 27 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_effect = model.gene_effect\n",
    "gene_effect = gene_effect.subtract(model.gene_effect.loc[:, neg_con].median(axis=1).median(), axis=0)\n",
    "gene_effect = gene_effect.divide(model.gene_effect.loc[:, pos_con].median(axis=1).abs().median(), axis=0)\n",
    "gene_effect"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1820bb23-d801-4895-8bd8-0648de072dc7",
   "metadata": {},
   "source": [
    "# Save outputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8ab789ee-edaa-4e0b-b40e-4aef614d95c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "lfc.to_csv(\"outputs/logfold_change_9p21_3.csv\")\n",
    "gene_effect.to_csv(\"outputs/gene_effect_9p21_3.csv\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
